NASA-IMPACT/csdap-cumulus
# Fix Terrascript build/deploy failures

Fixes a number of deployment problems:

- A misleading "Inappropriate ioctl for device" error, which was caused by asking Terraform to create a zip file from a directory containing symlinks. The symlinks were removed, as they were unnecessary.
- An error due to TypeScripts odd behavior when incrementally compiling. Incremental compiling was disabled, and the cleanup of compilation artifacts was also fixed to ensure proper cleanup. Improper/incomplete cleanup may have also been a contributing factor.
- Deployments that require more than an hour would fail due to Terraspace's default deploy timeout of 1 hour. This was wreaking havoc as it caused Terraform to lose touch with reality. Bumped timeout to 4 hours.
- Added "Terraform Doctor" script to easily fix "duplicate resource" errors that can be caused by abruptly killing Terraform (such as what would happen when Terraspace would time out).
- Added "nuke" script to allow for straightforward destruction of a Cumulus deployment.

To test this PR, do the following:

- Of course, checkout this branch first
- Run `make docker` to update your Docker image with the Terraspace timeout fix
- Run `make all-up-yes` to perform a full deployment

If you encounter any "duplicate resource" (or "already exists") errors during deployment, run the following command for the module that fails with such errors: `make terraform-doctor-MODULE`, where `MODULE` is the failing module (`cumulus`, `data-persistence` or `rds-cluster`). For example: `make terraform-doctor-cumulus` to fix "duplicate resource" errors produced during deployment of the `cumulus` module, as part of running `make all-up-yes`.

Please also proof-read the following docs:

- `docs/OPERATING.md`
- `docs/TROUBLESHOOTING.md`

You can also _partially_ test the "nuke" script WITHOUT nuking your stack by running `make nuke` and then simply pressing enter/return when prompted for confirmation (without typing the confirmation phrase that should be shown in the confirmation message, the operation will be aborted).

Fixes #33
